android/engine/src/main/java/org/smartregister/fhircore/engine/task/WorkflowCarePlanGenerator.kt

```diff
@@ -0,0 +1,322 @@
+
+@Singleton
+class WorkflowCarePlanGenerator
+@Inject
+constructor(
+  val knowledgeManager: KnowledgeManager,
+  val fhirOperator: FhirOperator,
+  val fhirEngine: FhirEngine,
+  @ApplicationContext val context: Context,
+) {
+
+  private var cqlLibraryIdList = ArrayList<String>()
+  private val jsonParser = FhirContext.forCached(FhirVersionEnum.R4).newJsonParser()
+
+  private fun writeToFile(resource: Resource): File {
+    val fileName =
+      if (resource is MetadataResource && resource.name != null) {
+        resource.name
+      } else {
+        resource.idElement.idPart
+      }
+    return File(context.filesDir, fileName).apply {
+      writeText(jsonParser.encodeResourceToString(resource))
+    }
+  }
+
+  /**
+   * Extracts resources present in PlanDefinition.contained field
+   *
+   * We cannot use $data-requirements on the [PlanDefinition] yet. So, we assume that all knowledge
+   * resources required to $apply a [PlanDefinition] are present within `PlanDefinition.contained`
+   *
+   * @param planDefinition PlanDefinition resource for which dependent resources are extracted
+   */
+  suspend fun getPlanDefinitionDependentResources(
+    planDefinition: PlanDefinition,
+  ): Collection<Resource> {
+    var bundleCollection: Collection<Resource> = mutableListOf()
+
+    for (resource in planDefinition.contained) {
+      resource.meta.lastUpdated = planDefinition.meta.lastUpdated
+      if (resource is Library) {
+        cqlLibraryIdList.add(IdType(resource.id).idPart)
+      }
+      knowledgeManager.install(writeToFile(resource))
+
+      bundleCollection += resource
+    }
+    return bundleCollection
+  }
+
+  /**
+   * Knowledge resources are loaded from [FhirEngine] and installed so that they may be used when
+   * running $apply on a [PlanDefinition]
+   */
+  private suspend fun loadPlanDefinitionResourcesFromDb() {
+    // Load Library resources
+    val availableCqlLibraries = fhirEngine.search<Library> {}
+    val availablePlanDefinitions = fhirEngine.search<PlanDefinition> {}
+    for (cqlLibrary in availableCqlLibraries) {
+      fhirOperator.loadLib(cqlLibrary)
+      knowledgeManager.install(writeToFile(cqlLibrary))
+      cqlLibraryIdList.add(IdType(cqlLibrary.id).idPart)
+    }
+    for (planDefinition in availablePlanDefinitions) {
+      getPlanDefinitionDependentResources(planDefinition)
+    }
+  }
+
+  /**
+   * Executes $apply on a [PlanDefinition] for a [Patient] and creates the request resources as per
+   * the proposed [CarePlan]
+   *
+   * @param planDefinitionId PlanDefinition resource ID for which $apply is run
+   * @param patient Patient resource for which the [PlanDefinition] $apply is run
+   * @param requestResourceConfigs List of configurations that need to be applied to the request
+   *   resources as a result of the proposed [CarePlan]
+   */
+  suspend fun applyPlanDefinitionOnPatient(
+    planDefinition: PlanDefinition,
+    patient: Patient,
+    data: Bundle = Bundle(),
```

Comment:
this doesn't appear to be used

Reply:
calling from here https://github.com/opensrp/fhircore/blob/c65cdbbabdcd397bbe790dd0514b128a765fa459/android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirCarePlanGenerator.kt#L122-L127

Reply:
ah missed that!